src/rgw/rgw_op.cc

```diff
@@ -2855,19 +2936,34 @@ void RGWPutObj::execute()
     dout(10) << "v4 auth ok" << dendl;
 
   }
-
   op_ret = store->check_quota(s->bucket_owner.get_id(), s->bucket,
                               user_quota, bucket_quota, s->obj_size);
   if (op_ret < 0) {
     ldout(s->cct, 20) << "second check_quota() returned op_ret=" << op_ret << dendl;
     goto done;
   }
 
-  if (need_calc_md5) {
-    processor->complete_hash(&hash);
-  }
   hash.Final(m);
 
+  if (compression_enabled && compressor->is_compressed()) {
+    bufferlist tmp;
+    RGWCompressionInfo cs_info;
+    cs_info.compression_type = s->cct->_conf->rgw_compression_type;
+    cs_info.orig_size = s->obj_size;
+    cs_info.blocks = move(compressor->get_compression_blocks());
+    ::encode(cs_info, tmp);
+    attrs[RGW_ATTR_COMPRESSION] = tmp;
+  }
+
+  // add attr to bucket to know original  size of data
+  res = rgw_bucket_compression_info_from_attrset(s->bucket_attrs, bucket_size);
+  if (res < 0)
+    lderr(s->cct) << "ERROR: failed to read decompressed bucket size, cannot update stat" << dendl;
+  op_ret = update_compressed_bucket_size(s->obj_size, bucket_size);
+  if (op_ret < 0)
+    ldout(s->cct, 0) << "NOTICE: put_bucket_info on bucket=" << s->bucket.name
+         << " returned err=" << op_ret << dendl;
+
```

Comment:
@aclamk i'm concerned about this extra request to update the bucket's size for every PUT

i made a quick hack that passes this `s->obj_size` directly through `RGWPutObjProcessor::complete()` -> `RGWRados::Object::Write::write_meta()` -> `RGWRados::Bucket::UpdateIndex::complete()` -> `cls_rgw_bucket_complete_op()`, so that the bucket index tracked this uncompressed size instead of the compressed size calculated by `RGWPutObjProcessor`. this was passing s3tests for me

i brought this up in the standup, and @yehudasa asked that we pass in both the compressed and uncompressed size, so the bucket stats could track the sum for each. i've made the necessary changes in cls/rgw and have been working my way up the chain. i hope to have something ready to show you today or tomorrow